docker/Dockerfile

```diff
@@ -1,41 +1,75 @@
-    truncate -s 0 /var/log/*log
+# build dependencies:
+# -virtual: create virtual package for later deletion
+# - build-base for alpine fundamentals
+# - ruby-dev/libc-dev for compiling raindrops, at least
+# - libxml2-dev/libxslt-dev for nokogiri, at least
+# - postgres-dev for pg/activerecord gems
+# - git for installing gems referred to use a git:// uri
+#
+# runtime dependencies:
+# - file: for paperclip file type spoofing check
+# - nodejs: for ExecJS and asset compilation
+# - runit for process management (because we have multiple services)
+# - libreoffice/pdftk: for pdf generation
+# - redis: for backend key-value store
+#
+RUN apk --no-cache add --virtual build-dependencies \
+                    build-base \
+                    libxml2-dev \
+                    libxslt-dev \
+                    postgresql-dev \
+                    git \
+&& apk --no-cache add \
+                  file \
+                  nodejs \
+                  linux-headers \
+                  runit \
+                  pdftk \
+                  libreoffice \
+                  redis
 
-# SSH proxy settings
-ENV SSH_AUTH_SOCK /tmp/ssh-auth
-ENV SSH_AUTH_PROXY_PORT 1234
-
-RUN mkdir -p /usr/src/app
-RUN bundle config --global without test:development
+# add non-root user and group with alpine first available uid, 1000
+RUN addgroup -g 1000 -S appgroup \
+&& adduser -u 1000 -S appuser -G appgroup
 
+# create app directory in conventional, existing dir /usr/src
+RUN mkdir -p /usr/src/app && mkdir -p /usr/src/app/tmp
 WORKDIR /usr/src/app
 
-COPY Gemfile /usr/src/app/
-COPY Gemfile.lock /usr/src/app/
+######################
+# DEPENDENCIES START #
+######################
+# Env vars needed for dependency install and asset precompilation
+ENV RAILS_ENV='production'
+
+COPY Gemfile* ./
 
-# Hack to install private gems
-#RUN socat UNIX-LISTEN:$SSH_AUTH_SOCK,fork TCP4:$(ip route|awk '/default/ {print $3}'):$SSH_AUTH_PROXY_PORT & bundle install
-RUN bundle install
+# only install production dependencies,
+# build nokogiri using libxml2-dev, libxslt-dev
+RUN gem install bundler -v 1.17.3 \
+&& bundle config --global without test:development \
+&& bundle config build.nokogiri --use-system-libraries \
+&& bundle install
 
-COPY . /usr/src/app
-RUN mkdir -p /usr/src/app/public/assets
+####################
+# DEPENDENCIES END #
+####################
+COPY . .
 
-COPY ./docker/run.sh $APP_HOME/run.sh
-RUN chmod 777 $APP_HOME/run.sh
+# precompile assets, to be owned by appuser
+RUN bundle exec rake assets:precompile SECRET_KEY_BASE=a-real-secret-key-is-not-needed-here
 
-EXPOSE 80
+# tidy up installation
+RUN apk del build-dependencies
 
-#SECRET_TOKEN set here because otherwise devise blows up during the precompile.
-RUN bundle exec rake assets:precompile RAILS_ENV=production SECRET_KEY_BASE=a-real-secret-key-is-not-needed-here
+# non-root/appuser should own only what they need to
+RUN chown -R appuser:appgroup log tmp db
 
-CMD "$APP_HOME/run.sh"
+USER appuser
```

Comment:
I was talking to @colinbruce about this (`USER <name>` vs `USER <uid>`) yesterday, and today came across this in the cloud platforms guidance which clarifies it
https://user-guide.cloud-platform.service.justice.gov.uk/tasks.html#how-to-adapt-to-the-pod-security-policies
`You must specify the user by its numeric UID`
Colin said this branch was deploying to live-1 ok, so I'm not sure, but it looks like you may have to go uid like we have?

Reply:
cool, cheers